### PR TITLE
PR for #1 and #2: top level YAML as Sequence (list)

### PR DESCRIPTION
**Pull request by justb4**
Fix for #1 and #2 plus documentation. All Shippable CI is passing for Python 3 and 2.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -36,15 +36,17 @@ Example:
...
       - user: dee
         name: Douglas
 
-    Group:
+    - Group:
       - name: Ramones
         members: [joey.profile, dee.profile]
 
--  The root of YAML contain ``mapper`` names i.e. ``User``
+-  The root of YAML contain ``mapper`` names e.g. ``User``, ``Profile`` etc
```
Comment (schettino72) [inline README.rst]: it is not a mapper. it is a list (as you said below). where an element is a mapping with a single element.

File: README.rst [hunk 2]
```diff
@@ -36,15 +36,17 @@ Example:
...
+-  The ``mapper`` names should be a ``list`` to enforce ordering e.g. ``- User``, ``- Profile``
+-  Order of ``mapper`` names should follow relationship (``__key__``) dependencies
```
Comment (schettino72) [inline README.rst]: at this point `__key__` was not introduced yet. IMO no need to mention this about ordering, lists are ordered...

File: setup.py [hunk 3]
```diff
@@ -16,11 +16,19 @@ def readme():
...
         'Intended Audience :: Developers',
         ],
       py_modules = ['sqla_yaml_fixtures'],
-      install_requires = ['SQLAlchemy', 'PyYAML'],
+      install_requires = [
+          'SQLAlchemy',
+          'PyYAML'
+      ],
+      extras_require={
+          ':python_version=="2.7"': ['backports.functools_lru_cache'],
```
Comment (schettino72) [inline setup.py:60]: Nice you used the modern way. But most people have a `pip` version that does not support it. I am ok to using this but a clear error message saying "upgrade pip" should be given for old versions of pip.

File: sqla_yaml_fixtures.py [hunk 4]
```diff
@@ -1,13 +1,19 @@
 __version__ = (0, 1, 2)
 
-import functools
+try:
+    # Python 3
+    from functools import lru_cache
+except ImportError:
```
Comment (schettino72) [inline sqla_yaml_fixtures.py]: can you add a `# pragma: no cover` for python 2 specific code. This helps target 100% test coverage.

File: sqla_yaml_fixtures.py [hunk 5]
```diff
@@ -1,13 +1,19 @@
...
+    # For Python 2: 
+    # use https://pypi.python.org/pypi/backports.functools_lru_cache
```
Comment (schettino72) [inline sqla_yaml_fixtures.py]: the package information is already `setup.py` so i dont see why it should be mentioned here.

File: sqla_yaml_fixtures.py [hunk 6]
```diff
@@ -1,13 +1,19 @@
...
+    # pip install backports.functools_lru_cache
+    from backports.functools_lru_cache import lru_cache
 
 import yaml
 import sqlalchemy
 from sqlalchemy.orm.relationships import RelationshipProperty
 
 
-
```
Comment (schettino72) [inline sqla_yaml_fixtures.py]: It is on PEP8 3 blank lines to separate imports 
I sometimes use more blank lines :grin:

File: sqla_yaml_fixtures.py [hunk 7]
```diff
@@ -123,9 +129,16 @@ def load(ModelBase, session, fixture_text):
...
+    # to enforce that FKs (__key__ entries) are defined first
     data = yaml.load(fixture_text)
+    if not isinstance(data, list):
+        raise ValueError('top level YAML should be sequence (list)')
+
     store = Store()
-    for model_name, instances in data.items():
+    for model_entry in data:
+        # Model entry is dict with one item
+        model_name, instances = model_entry.popitem()
```
Comment (schettino72) [inline sqla_yaml_fixtures.py]: Better check and raise an error if the dict contains more than one entry.

**Review by schettino72 — CHANGES_REQUESTED** [review]
First thanks for working on this.

* need to update CHANGES file
* maybe we should add an AUTHORS file
* since we are adding python2, i guess the code should just work on python3.4 and 3.5. just include it in shippable.yaml and setup.py

File: README.rst (hunk 1 above)
Comment (justb4) [reply README.rst]: You're right (thought of SQLAlchemy mappers), will change

File: README.rst (hunk 2 above)
Comment (justb4) [reply README.rst]: ok

File: setup.py (hunk 3 above)
Comment (justb4) [reply setup.py:60]: Struggled long with this one using ``install_requires`` but somehow could not get right expression for something like `< python_version < "3.0"`...where to put error msg?

File: sqla_yaml_fixtures.py (hunk 4 above)
Comment (justb4) [reply sqla_yaml_fixtures.py]: Ok.

File: sqla_yaml_fixtures.py (hunk 7 above)
Comment (justb4) [reply sqla_yaml_fixtures.py]: ok, you're thorough 👍

File: sqla_yaml_fixtures.py (hunk 6 above)
Comment (justb4) [reply sqla_yaml_fixtures.py]: ok, I use `flake8` in other projects in CI. Thought it was included If I run it I get this warning and many others:

```
./sqla_yaml_fixtures.py:7:20: W291 trailing whitespace
./sqla_yaml_fixtures.py:12:1: E402 module level import not at top of file
./sqla_yaml_fixtures.py:13:1: E402 module level import not at top of file
./sqla_yaml_fixtures.py:14:1: E402 module level import not at top of file
./sqla_yaml_fixtures.py:44:5: E265 block comment should start with '# '
./sqla_yaml_fixtures.py:50:11: E271 multiple spaces after keyword
./sqla_yaml_fixtures.py:56:1: E303 too many blank lines (3)
./sqla_yaml_fixtures.py:72:14: E261 at least two spaces before inline comment
./sqla_yaml_fixtures.py:99:25: E261 at least two spaces before inline comment
./test_sqla_yaml_fixtures.py:16:1: E302 expected 2 blank lines, found 1
./test_sqla_yaml_fixtures.py:33:80: E501 line too long (80 > 79 characters)
./test_sqla_yaml_fixtures.py:71:1: E303 too many blank lines (3)
./test_sqla_yaml_fixtures.py:75:1: E302 expected 2 blank lines, found 3
./test_sqla_yaml_fixtures.py:79:1: E302 expected 2 blank lines, found 1
./test_sqla_yaml_fixtures.py:85:1: E302 expected 2 blank lines, found 1
./test_sqla_yaml_fixtures.py:130:1: E303 too many blank lines (3)
./test_sqla_yaml_fixtures.py:209:1: E303 too many blank lines (3)
./test_sqla_yaml_fixtures.py:260:1: W391 blank line at end of file

```

File: README.rst (hunk 1 above)
Comment (schettino72) [reply README.rst]: oh, yes. i was referring to SQLAlchemy mappers... and I just wrote this a week ago :disappointed:

What about

The root of YAML contains a list of SQLAlchemy mapper names

File: sqla_yaml_fixtures.py (hunk 6 above)
Comment (schettino72) [reply sqla_yaml_fixtures.py]: Sorry, I double check and could not find anything about 3 lines after import. It seems I made that up...

While I kind of mostly follow pep8, personally I dont like this so strict blank line rules (and the checker is more strict then PEP8 itself).

First, because I use blank lines with a semantic meaning not fixed rules.
Second because other stuff I think it is not worth the trouble. 

The module level warning also seems wrong because the only thing before imports is `__version__`. 

So I would prefer not introduce this tool... I wont complain about how many blank lines you use. I just mentioned here because you made a change in area that otherwise had no changes.

File: sqla_yaml_fixtures.py (hunk 7 above)
Comment (justb4) [reply sqla_yaml_fixtures.py]: Propose to also skip zero length model entries. I have a prog that generates a `sqla_fixture` from SQLAlchemy model. If the table is empty it generates just an empty `- Name:`. Ok?

File: sqla_yaml_fixtures.py (hunk 6 above)
Comment (justb4) [reply sqla_yaml_fixtures.py]: hmm, thought you were PEP8-thorough, so just made changes so we get no flake8 warnings....just clean Python, I also saw the warnings in my IDE (IntelliJ). But no need for `flake8`, it can be very annoying as well, as the CI may fail on a single space on a blank line...

File: README.rst (hunk 1 above)
Comment (justb4) [reply README.rst]: ok

Comment (justb4) [reply README.rst]: The official YAML terminology is `sequence` (`list`) and `mapping` (`dict`) hence my confusion.

File: sqla_yaml_fixtures.py (hunk 6 above)
Comment (schettino72) [reply sqla_yaml_fixtures.py]: if you already fixed it. great. apply the fixes.

File: README.rst (hunk 1 above)
Comment (justb4) [reply README.rst]: So *The root of YAML contains a sequence of SQLAlchemy mapper names*

File: sqla_yaml_fixtures.py (hunk 7 above)
Comment (schettino72) [reply sqla_yaml_fixtures.py]: I guess empty is ok. it is not something like will surprise an user with a silent failure.

File: README.rst (hunk 1 above)
Comment (justb4) [reply README.rst]: I now have:

-  The root of YAML contains a *sequence* of ``mapper names`` e.g. ``- User``, ``- Profile`` etc
-  The order of these names should follow relationship dependencies
-  Every name should contain a *sequence* of instances
-  Each instance is a *mapping* of *attribute* -> *value*
-  the attributes are taken from the mapper ``__init__()`` (usually an
   attributes maps to a column)
-  The special field ``__key__`` can be used to identify this instnace
   in a relationship reference e.g. The ``Profile.user``
-  Note that any ``__key__`` MUST be globally **unique**
-  In a *to-one* relationship the data can be directly nested in the
   parent data definition
-  References can access attributes using a *dot* notation, e.g.
   ``joey.profile``
-  *to-many* relationships can be added as a list of references

File: setup.py (hunk 3 above)
Comment (schettino72) [reply setup.py:60]: I mean something like this:
https://github.com/ipython/ipython/blob/master/setup.py#L34

File: sqla_yaml_fixtures.py (hunk 7 above)
Comment (justb4) [reply sqla_yaml_fixtures.py]: Looking ahead: a generic `sql_yaml_fixtures.dump()` would be very useful for e.g. a generic `import/export` facility in e.g. a Flask app. But this is a separate feature maybe for later...

**Comment by justb4** [comment]
will modify CHANGES (was there a 0.1.0, .. ok yes I see release) and add AUTHORS. #1 and #2 will go in 0.1.2 I guess?

**Comment by schettino72** [comment]
> will modify CHANGES (was there a 0.1.0?) and add AUTHORS. #1 and #2 will go in 0.1.2 I guess?

1.0 and 1.1 there were problems with the long_description but no code changes, thats why i made new releases but did not bother to update CHANGES.

Next release should be 0.2.0 because it is not compatible with 0.1.x

And since we are now 2 users i guess on setup.py we could change *Development status* from *alpha* to * beta* :smiley:

File: sqla_yaml_fixtures.py (hunk 7 above)
Comment (justb4) [reply sqla_yaml_fixtures.py]: Also added test for empty case...

**Comment by justb4** [comment]
Ok for 0.2.0 and beta.

File: setup.py (hunk 3 above)
Comment (justb4) [reply setup.py:60]: Ok, I propose this:

```
from __future__ import print_function
from setuptools import setup
import sys

# Assume pip is new enough
pip_message = None
try:
    import pip
    pip_version = tuple([int(x) for x in pip.__version__.split('.')[:3]])
    if pip_version < (9, 0, 1) :
        pip_message = 'Your pip version is out of date, please install pip >= 9.0.1. '\
        'pip {} detected.'.format(pip.__version__)
        
except Exception:
    pip_message = 'This may be due to an out of date pip. Make sure you have pip >= 9.0.1.'


if pip_message:
    # Some error from pip version check
    error = """
    Python {py} detected.
    Exiting setup: 
    {pip}
    """.format(py=sys.version_info, pip=pip_message )

    print(error, file=sys.stderr)
    sys.exit(1)
```

File: README.rst [hunk 8]
```diff
@@ -36,29 +36,30 @@ Example:
...
 -  Note that any ``__key__`` MUST be globally **unique**
 -  In a *to-one* relationship the data can be directly nested in the
    parent data definition
--  References can access attributes using a *dot* notaion, i.e.
+-  References can access attributes using a *dot* notation, e.g.
    ``joey.profile``
 -  *to-many* relationships can be added as a list of references
 
-The mapper definition for this example is in the `test file`_.
+The mapper definition for this example is in the [test file](test_sqla_yaml_fixtures.py).
```
Comment (schettino72) [inline README.rst]: This is a ReST file not markdown. Check the bottom of the file, you will see where this reference points to.

File: CHANGES [hunk 9]
```diff
@@ -1,4 +1,15 @@
...
-- Initial release
\ No newline at end of file
+- Initial release
+
+0.1.1 (2017-06-14)
+==================
+
+- convert README to ReST
+
+0.2.0 (2017-06-??)
```
Comment (schettino72) [inline CHANGES]: Usually releases are listed in reverse chronological order. So latest release on top.

File: AUTHORS [hunk 10]
```diff
@@ -0,0 +1,8 @@
+sqla_yaml_fixtures was originally created in june 2017 by Eduardo Naufel Schettino.
+
+Here is an inevitably incomplete list of MUCH-APPRECIATED CONTRIBUTORS --
```
Comment (schettino72) [inline AUTHORS:3]: :laughing:

File: README.rst (hunk 8 above)
Comment (justb4) [reply README.rst]: aha, assumed README.md (most on GH/Gitter follows `.md` format) .

Comment (justb4) [reply README.rst]: Then this style will be more portable among forks/branches: 

```
`test file <test_sqla_yaml_fixtures.py>`_
```

Comment (justb4) [reply README.rst]: forget last: I now see the `README.rst` goes all the way to `PyPi` so yes need absolute URL and thus 
```
`test file`_
```

File: CHANGES (hunk 9 above)
Comment (justb4) [reply CHANGES]: Reversed, you still need to set a date for 0.2.0.

File: AUTHORS (hunk 10 above)
Comment (justb4) [reply AUTHORS:3]: For these thousands of contributors we expect in future ;-).

**Comment by schettino72** [comment]
Great. thanks for you patience :) Please go ahead and merge it.
 
And lets do another release... do you have a pypi account? I will add you as an admin for the project there also.

**Comment by justb4** [comment]
Ok, done, good to work with a real thorough CI-dev :-), yes also on PyPi: `justb4`

**Comment by schettino72** [comment]
I added you on pypi. not sure you get an email or any kind of notification. feel free to push a release :rocket:

**Comment by justb4** [comment]
no notification, but I see I've been added.

**Comment by schettino72** [comment]
can i do a release? or you want to put anything else before releasing it?

**Comment by justb4** [comment]
Ok with me to release.
